core/types/receipt.go

```diff
@@ -425,6 +426,12 @@ func (rs Receipts) DeriveFields(config *params.ChainConfig, hash common.Hash, nu
 		} else {
 			rs[i].GasUsed = rs[i].CumulativeGasUsed - rs[i-1].CumulativeGasUsed
 		}
+		// The effective gas price can be calculated based on baseFee for EIP1559 transactions
+		if txs[i].Type() == 2 {
+			rs[i].EffectiveGasPrice = new(big.Int).Add(baseFee, txs[i].EffectiveGasTipValue(baseFee))
+		} else {
+			rs[i].EffectiveGasPrice = new(big.Int).Set(txs[i].GasPrice())
+		}
```

Comment:
I think this logic will bitrot if we leave it here (as soon as we add a new tx type). I'd rather have a method 
```
func (tx *Transaction) EffectiveGas(baseFee *big.Int) (*big.Int, error) 
```
And that method should use the methods that the transaction types expose. If we do it this way, we can make it so any new type will be 'forced' to handle it